kie-server-parent/kie-server-services/kie-server-services-prometheus/src/main/java/org/kie/server/services/prometheus/PrometheusKieServerExtension.java

```diff
@@ -59,13 +72,62 @@ public boolean isInitialized() {
 
     @Override
     public boolean isActive() {
-        return !disabled;
+        return disabled == false;
     }
 
     @Override
     public void init(KieServerImpl kieServer, KieServerRegistry registry) {
         this.context = registry;
+
+        registerDefaultDescriptor();
+
+        //Prometheus Monitoring
+        KieServerExtension jBPMExtension = context.getServerExtension(JbpmKieServerExtension.EXTENSION_NAME);
+        if (jBPMExtension != null) {
+
+            final KModuleDeploymentService deploymentService = jBPMExtension.getAppComponents(KModuleDeploymentService.class);
+            if (deploymentService != null) {
+                deploymentService.addListener(new PrometheusDeploymentEventListener());
+            }
+
+            final ExecutorServiceImpl executorService = jBPMExtension.getAppComponents(ExecutorServiceImpl.class);
+            if (executorService != null) {
+                executorService.addAsyncJobListener(new PrometheusJobListener());
+            }
+
+            final QueryServiceImpl queryService = jBPMExtension.getAppComponents(QueryServiceImpl.class);
+            if (queryService != null) {
+                final DataSetDefRegistry dataSetDefRegistry = queryService.getDataSetDefRegistry();
+                final PrometheusDataSetListener listener = new PrometheusDataSetListener(dataSetDefRegistry);
+                listener.init();
+                dataSetDefRegistry.addListener(listener);
+            }
+
+            final RuntimeDataService dataService = jBPMExtension.getAppComponents(RuntimeDataService.class);
+            if (dataService != null) {
+                final Collection<ProcessInstanceDesc> processInstances = dataService.getProcessInstances(
+                        asList(ProcessInstance.STATE_ACTIVE), null, new QueryContext(0, Integer.MAX_VALUE));
+                processInstances.forEach(pi -> recordRunningProcessInstance(pi.getDeploymentId(), pi.getProcessId()));
+            }
+        }
+
+        KieServerExtension caseExtension = context.getServerExtension(CaseKieServerExtension.EXTENSION_NAME);
+        if (caseExtension != null) {
+            final CaseManagementRuntimeDataServiceBase caseRuntime = caseExtension.getAppComponents(CaseManagementRuntimeDataServiceBase.class);
+            if (caseRuntime != null) {
+                final CaseInstanceList caseInstances = caseRuntime.getCaseInstances(asList(CaseStatus.OPEN.getName()), 0, Integer.MAX_VALUE, null, false);
+                for (CaseInstance instance : caseInstances.getCaseInstances()) {
+                    recordRunningCaseInstance(instance.getCaseDefinitionId());
+                }
+            }
+        }
+
         initialized = true;
+        LOGGER.info("{} started", toString());
+    }
+
+    protected void registerDefaultDescriptor() {
+        System.setProperty("org.kie.deployment.desc.location", "classpath:/META-INF/prometheus-deployment-descriptor-defaults.xml");
```

Comment:
@MarianMacik correct, they would need to redefine the Prometheus ones in their custom descriptor.

Reply:
Well, the thing is that with this approach even if they redefine it with a custom descriptor, their descriptor won't be taken into account since it will be overridden by this property setter.

I discussed this with Karel and instead of setting this via the property, we can use a similar approach as for CaseMgmt. There is this piece of code:
https://github.com/kiegroup/droolsjbpm-integration/blob/83e57e9ebadc2ab6929977bc3d496411ccf503cf/kie-server-parent/kie-server-services/kie-server-services-case-mgmt/src/main/java/org/kie/server/services/casemgmt/CaseKieServerExtension.java#L158-L162

which registers `CaseConfigurationDeploymentListener` only when case mgmt extension is enabled. That listener then reads deployment descriptor and adds CaseEventListeners to CaseEventSupport as you can see here:

https://github.com/kiegroup/jbpm/blob/ec824f1bf1889da57cb4d6368ad6faa73fefc21f/jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/event/CaseConfigurationDeploymentListener.java#L69

So we can do something similar. We can just create `PrometheusConfigurationDeploymentListener` which will just do something like:

```
public void onDeploy(DeploymentEvent event) {
        InternalRuntimeManager runtimeManager = (InternalRuntimeManager) event.getDeployedUnit().getRuntimeManager();
        RuntimeEngine runtimeEngine =  runtimeManager.getRuntimeEngine(EmptyContext.get());
        runtimeEngine.getKieSession().addEventListener(new PrometheusProcessEventListener());
        ((EventService<TaskLifeCycleEventListener>)runtimeEngine.getTaskService()).registerTaskEventListener(new PrometheusTaskEventListener());
        if (runtimeManager instanceof PerCaseRuntimeManager) {
            ((PerCaseRuntimeManager) runtimeManager).getCaseEventSupport().addEventListener(new PrometheusCaseEventListener());
        }
```

@mswiderski wdyt? Will it work? I think this is less intrusive (although it is more code) than to setting global deployment descriptor via system property.

I especially don't know if `runtimeManager.getRuntimeEngine(EmptyContext.get());` is enough in case there will be PPI or PerRequest runtime manager.

Reply:
OK, I had a discussion with @mswiderski and we should use the deployment descriptor way, but we should first check if a user has already defined one with a system property, so this extension doesn't override it.

Reply:
ok, I tried the approach you mentioned, it works fine for case management, but I couldn't find a way to register the process and task listeners.

Reply:
will add a warning if system property already defined.

Reply:
Thanks, looks good now!